python/ray/serve/tests/test_warnings.py

```diff
@@ -0,0 +1,64 @@
+import time
+
+from ray import serve
+from ray.serve.backend_state import (SLOW_STARTUP_WARNING_S,
+                                     SLOW_STARTUP_WARNING_PERIOD_S)
+
+
+def test_slow_allocation_warning(serve_instance, capsys):
+    # this deployment can never be scheduled
+    @serve.deployment(ray_actor_options={"num_cpus": 99999})
+    class D:
+        def __init__(self):
+            pass
+
+    num_replicas = 2
+    D.options(num_replicas=num_replicas).deploy(_blocking=False)
+
+    expected_warning = (f"Deployment '{D.name}' has "
+                        f"{num_replicas} replicas that have taken "
+                        f"more than {SLOW_STARTUP_WARNING_S}s "
+                        f"to be scheduled.")
+
+    # wait long enough for the warning to be printed
+    # with a small grace period
+    time.sleep(SLOW_STARTUP_WARNING_PERIOD_S * 1.5)
```

Comment:
Hmm, it's a bit tricky because this behaviour is purely time-based. Maybe we could re-use `MockTimer` from `tests/test_backend_state.py` for this?

Reply:
update: AFAIK we can't use MockTimer since the serve controller runs in a different process from the test, as to my knowledge you cannot mock in another process.